File: python/basics/stack/monotonous_stack/739_daily_temperatures.py
#https://leetcode.com/problems/daily-temperatures/description/

#monotonous stack
#O(n)
#space O(n)
def daily_temperatures_stack(temperatures):
    result =[0]*len(temperatures)
    stk = []
    for i in range(0,len(temperatures)):
        while stk and temperatures[i] > stk[len(stk)-1][0]:
            (t, index) = stk.pop()
            result[index] = i - index
        stk.append((temperatures[i], i))
    return result


#O(n**n)
def daily_temperatures_bruteforce(temperatures):
    result =[0]*len(temperatures)
    for i in range(len(temperatures)):
        for j in range(i+1,len(temperatures)):
            if temperatures[i] < temperatures[j]:
                result[i] = j-i
                break
    return result


temperatures = [73,74,75,71,69,72,76,73]
result = daily_temperatures_bruteforce(temperatures)
print("result-1:" + str(result))

result = daily_temperatures_stack(temperatures)
print("result-2:" + str(result))


#Output: [1,1,4,2,1,1,0,0]

temperatures = [30,40,50,60]
result = daily_temperatures_bruteforce(temperatures)
print("result-1:" + str(result))

result = daily_temperatures_stack(temperatures)
print("result-2:" + str(result))


#Output: [1,1,1,0]

temperatures = [30,60,90]
result = daily_temperatures_bruteforce(temperatures)
print("result-1:" + str(result))

result = daily_temperatures_stack(temperatures)
print("result-2:" + str(result))

temperatures = [1,3,2,4]
result = daily_temperatures_stack(temperatures)
print("result-2:" + str(result))


#Output: [1,1,0]

File: problem_solving/stack/cpp/monotonic_stack/739_daily_temperatures.cpp
/*
#https://leetcode.com/problems/daily-temperatures/description/

#monotonous stack
#O(n)
#space O(n)
*/

#include "../../../../cpp/common_header.h"
#include "../../../../cpp/utils/utils.h"
#include<stack>


std::vector<int> daily_temperatures_stack(std::vector<int> input) {
    int len = input.size();
    std::vector<int> results(len, -1);
    std::stack<int> st;
    int index = 0;
    for(int i = len-1; i >= 0; --i) {
        while(!st.empty() && input[st.top()] <= input[i]) {
            st.pop();
        }
        if(!st.empty()) {
            results[i] = st.top()-i;
        }
        st.push(i);
    }
    return results;
}


int main() {
    std::vector<int> temperatures = {73,74,75,71,69,72,76,73};
    std::vector<int> result = daily_temperatures_stack(temperatures);
    print_vector(result);
    std::cout << "result done " << std::endl;
    //#Output: [1,1,4,2,1,1,0,0]

    temperatures = {30,40,50,60};
    result = daily_temperatures_stack(temperatures);
    print_vector(result);
    std::cout << "result done " << std::endl;
    //#Output: [1,1,1,0]

    temperatures = {30,60,90};
    result = daily_temperatures_stack(temperatures);
    print_vector(result);
    std::cout << "result done " << std::endl;
    //1,1,-1,

    temperatures = {1,3,2,4};
    result = daily_temperatures_stack(temperatures);
    print_vector(result);
    std::cout << "result done " << std::endl;
    //#Output: [1,2,1,-1]
}